### DISCUSSION TITLE
How do I add a configuration file to an Axmol project?

### DISCUSSION THREAD

**Discussion by FHBlah** [axmolengine/axmol#1832] [Q&A] [answered]
Hi folks, I've successfully migrated my Cocos2d-x project to Axmol and it works beautifully. Next, I'd like to create a free version of the app that will be monetized with ads. As Axmol is a C++ engine, I've decided to use the [Firebase Google Mobile Ads C++ SDK](https://firebase.google.com/docs/admob/cpp/quick-start) (This is my first time incorporating ads in a project so if there is a better way, please let me know). I'm following the instructions in the **readme.md** file of the provided [test app](https://github.com/firebase/quickstart-cpp/blob/main/gma/testapp/readme.md). The instructions had me [register my iOS app with Firebase](https://firebase.google.com/docs/ios/setup). As a result of these steps, I've downloaded a configuration file called **GoogleService-Info.plist** which I am instructed to "_move into the root of my Xcode project_". This is where things get blurry.

Podfile installations have turned my Xcode project into an Xcode workspace. Is the "_root of my project_" the root of the Axmol project, is it the **build_ios_arm64** folder, or some other folder like **proj.ios_mac**? My **Podfile** is in the Axmol root but its first line is `project './build_ios_arm64/AdmobTest2.xcodeproj'` which tells CocoaPods where to find the actual Xcode project.

Also, dragging and dropping **GoogleService-Info.plist** into my project next to **Info.plist** won't work (I know, I've tried). The sample Firebase code compiles and runs but I can tell it can't find the config file because, in the output window, I get messages like the following:

```
2024-04-17 21:03:10.333927-0400 AdmobTest2[29247:100339287] 8.10.0 - [Firebase/Core][I-COR000005] No app has been configured yet.
2024-04-17 21:03:10.334078-0400 AdmobTest2[29247:100339287] 8.10.0 - [Firebase/Core][I-COR000003] The default Firebase app has not yet been configured. Add `FirebaseApp.configure()` to your application initialization. This can be done in in the App Delegate's application(_:didFinishLaunchingWithOptions:)` (or the `@main` struct's initializer in SwiftUI). Read more: https://goo.gl/ctyzm8.
2024-04-17 21:03:10.383505-0400 AdmobTest2[29247:100339304] 8.10.0 - [Firebase/Analytics][I-ACS025028] AdMob and Firebase are in use in this app, but are not linked. For analytics data consistency and integrity, it is highly recommended that you link your AdMob app to the corresponding Firebase app and re-download the GoogleServices-Info.plist file. See https://support.google.com/admob/answer/6383165 for more information.
```

In **CMakeLists.txt**, **Info.plist** is handled by specific instructions like:

```
       set(APP_UI_RES
            proj.ios_mac/mac/Icon.icns
            proj.ios_mac/mac/Info.plist
            )
```
and
```
set_target_properties(${APP_NAME} PROPERTIES MACOSX_BUNDLE_INFO_PLIST "${CMAKE_CURRENT_SOURCE_DIR}/proj.ios_mac/ios/targets/ios/Info.plist")
```

Should I be doing something similar for **GoogleServices-Info.plist**? Any insight on how to proceed would be greatly appreciated. Thanks.

**Comment by rh101** [comment, marked as answer]
Have you tried adding the `GoogleService-Info.plist` into the `<project>/Contents` folder? Once you do that, re-generate the project (via `axmol build ... -c -f` or whatever cmake command you used to generate it etc.).  For the `axmol build` command, make sure you pass `-f` to it in order to force it to regenerate the project, since cmake may not automatically detect changes in the `Contents` folder.

The file will then be treated as a resource that is copied across to the output bundle.  When Firebase runs, it'll find the file.  I know this works because I'm using it too, and there should be nothing else required to get it working.

After this re-run `pod install`, then open up the workspace.

To verify that it the file is referenced by the project correctly, go to `Build Phases` for `AdmobTest2` project. and you should find an entry for copying `GoogleServices-Info.plist` somewhere in there.

**Reply by FHBlah** [reply]
Thanks for responding, RH. Sorry I couldn't do better.

So, putting the config file in the `<project>/Contents` worked as you described. I checked `Build Phases` and confirmed the file was present. When I ran the app, I got the same error messages as before. I tried removing the file from the project and running again. This time I got these messages in the output window:
```
2024-04-19 14:50:24.456329-0400 AdmobTest2[35958:100854133] ERROR: Failed to read Firebase options from the app's resources (). Either make sure GoogleService-Info.plist is included in your build or specify options explicitly.
2024-04-19 14:50:24.456851-0400 AdmobTest2[35958:100854345] 8.10.0 - [Firebase/Core][I-COR000012] Could not locate configuration file: 'GoogleService-Info.plist'.
2024-04-19 14:50:24.462263-0400 AdmobTest2[35958:100854133] ERROR: Failed to read Firebase options from the app's resources (). Either make sure GoogleService-Info.plist is included in your build or specify options explicitly.
2024-04-19 14:50:24.462325-0400 AdmobTest2[35958:100854133] ERROR: App ID, API key, and Project ID must be specified in App options.
2024-04-19 14:50:24.462411-0400 AdmobTest2[35958:100854345] 8.10.0 - [Firebase/Core][I-COR000005] No app has been configured yet.
2024-04-19 14:50:24.462880-0400 AdmobTest2[35958:100854345] 8.10.0 - [Firebase/Core][I-COR000003] The default Firebase app has not yet been configured. Add `FirebaseApp.configure()` to your application initialization. This can be done in in the App Delegate's application(_:didFinishLaunchingWithOptions:)` (or the `@main` struct's initializer in SwiftUI). Read more: https://goo.gl/ctyzm8.
```

By the looks of it, my config file had been active the whole time. I was just misinterpreting the error messages. Clearly, there's a problem with my configuration. I'll look into it further. Thanks again for your help, at least now I know what the problem isn't.

**Reply by rh101** [reply]
Are you still missing the configuration call? The instructions on the [Firebase setup](https://firebase.google.com/docs/ios/setup) page specify this.

```
// Use Firebase library to configure APIs
[FIRApp configure];
```

If you haven't added it yet, then add it in `AppController.mm`, at the top of the method `didFinishLaunchingWithOptions`.

**Reply by FHBlah** [reply]
Thanks for the suggestion. I had noticed the error message instruction to add `[FIRApp configure]` or `FirebaseApp.configure()`to my `didFinishLaunchingWithOptions` funciton but since one was for Objective-C and the other was for Swift, I assumed they applied to the non-C++ SDKs. For one thing, I couldn't find an equivalent function call in the C++ SDK [Test App](https://github.com/firebase/quickstart-cpp/tree/main/gma/testapp). However, the Test App instructions **do** refer users to the same Firebase setup page as you did in your response. And Step 5. **does** instruct on how to "Initialize Firebase in your app". Again, there are only Objective-C and Swift examples provided but when rh101 makes a suggestion, one should pay attention.

So, in the `AppController.mm` file in the ```proj.ios_mac/ios/``` folder, as the first line the of the `didFinishLaunchingWithOptions` function, I added `[FIRApp configure];`. I was met with a `Use of undeclared identifier 'FIRApp'` message so I tried adding the suggested code snippet:
```
@import FirebaseCore;
@import FirebaseFirestore;
@import FirebaseAuth;

```
and got the following error:
`Use of '@import' when C++ modules are disabled, consider using -fmodules and -fcxx-modules`
I got rid of the import statements and tried using an #import statement like so:
`#import "FirebaseCore.h"`
and the compiler doesn't seem to mind it. On running the app, I no longer get the `[I-COR000003] The default Firebase app has not yet been configured` messages:
```
2024-04-20 15:05:24.532501-0400 AdmobTest2[39346:101096907] Loading UIApplication category for Firebase App
2024-04-20 15:05:24.578159-0400 AdmobTest2[39346:101097124] 8.10.0 - [Firebase/Core][I-COR000005] No app has been configured yet.
2024-04-20 15:05:24.647271-0400 AdmobTest2[39346:101097120] 8.10.0 - [Firebase/Analytics][I-ACS025028] AdMob and Firebase are in use in this app, but are not linked. For analytics data consistency and integrity, it is highly recommended that you link your AdMob app to the corresponding Firebase app and re-download the GoogleServices-Info.plist file. See https://support.google.com/admob/answer/6383165 for more information.
2024-04-20 15:05:24.741171-0400 AdmobTest2[39346:101097125] 8.10.0 - [Firebase/Analytics][I-ACS023007] Analytics v.8.10.0 started
2024-04-20 15:05:24.741842-0400 AdmobTest2[39346:101097125] 8.10.0 - [Firebase/Analytics][I-ACS023008] To enable debug logging set the following application argument: -FIRAnalyticsDebugEnabled (see http://goo.gl/RfcP7r)
2024-04-20 15:05:24.906474-0400 AdmobTest2[39346:101097124] 8.10.0 - [Firebase/Analytics][I-ACS800023] No pending snapshot to activate. SDK name: app_measurement
2024-04-20 15:05:24.914379-0400 AdmobTest2[39346:101097124] 8.10.0 - [Firebase/Analytics][I-ACS023012] Analytics collection enabled
2024-04-20 15:05:24.914465-0400 AdmobTest2[39346:101097124] 8.10.0 - [Firebase/Analytics][I-ACS023220] Analytics screen reporting is enabled. Call +[FIRAnalytics logEventWithName:FIREventScreenView parameters:] to log a screen view event. To disable automatic screen reporting, set the flag FirebaseAutomaticScreenReportingEnabled to NO (boolean) in the Info.plist
```

There's still a configuration problem I have to track down. Though the code runs, the initialization never completes. The code below never exits the while loop:

```
    App* app = ::firebase::App::Create();
    firebase::InitResult result;
    Future<AdapterInitializationStatus> future =
      firebase::gma::Initialize(*app, &result);
    
    while (future.status() == firebase::kFutureStatusPending)
    {
      // Initialization on-going, continue to wait.
    }
```
But I feel like I'm getting closer. I'll go over everything again tomorrow. Thanks again, RH.

**Reply by rh101** [reply]
> Again, there are only Objective-C and Swift examples provided but when rh101 makes a suggestion, one should pay attention.

That made me laugh! Now, on a more serious note, any Objective-C snippets simply means you need to add the calls in a `.m` or `.mm` file, which are already supported in your Axmol project.  

Also, you are correct, `#import "filename.h"` is what you use in Objective-C files.

Now, with regards to the `firebase::gma` namespace, I'm not entirely sure what is going on there.  For my project, I only make use of Crashlytics and Analytics, so this is working for me:

```
#include "firebase/analytics.h"
#include "firebase/app.h"

#if defined(__ANDROID__)
#include "platform/android/jni/JniHelper.h"
#endif

namespace myapp::firebase {
void initialize()
{
#if defined(__ANDROID__)
    auto appInstance = ::firebase::App::Create(ax::JniHelper::getEnv(), ax::JniHelper::getActivity());
#else
    auto appInstance = ::firebase::App::Create();
#endif
    if (appInstance)
        ::firebase::analytics::Initialize(*appInstance);
}

void terminate()
{
    ::firebase::analytics::Terminate();
}
}
```

I call the `initialize()` method from AppDelegate, for example:

```
bool AppDelegate::applicationDidFinishLaunching() 
{
    myapp::firebase::initialize();
...
}
```

In the test app gma example, [https://github.com/firebase/quickstart-cpp/blob/main/gma/testapp/src/common_main.cc](https://github.com/firebase/quickstart-cpp/blob/main/gma/testapp/src/common_main.cc), this is the code they use:
```
extern "C" int common_main(int argc, const char *argv[]) {
  firebase::App *app;
  LogMessage("Initializing Firebase App.");

  app = ::firebase::App::Create();

  LogMessage("Created the Firebase App %x.",
             static_cast<int>(reinterpret_cast<intptr_t>(app)));

  LogMessage("Initializing the GMA with Firebase API.");
  firebase::gma::Initialize(*app);

  WaitForFutureCompletion(firebase::gma::InitializeLastResult());
  if (firebase::gma::InitializeLastResult().error() !=
      firebase::gma::kAdErrorCodeNone) {
    // Initialization Failure. The error was already logged in
    // WaitForFutureCompletion, so simply exit here.
    return -1;
  }
...
}
```

So, perhaps you shouldn't use a `while` loop, and instead use an `if` condition, and actually check what error is being returned by `firebase::gma::InitializeLastResult().error()`; that should point you in the right direction.  When you do find out which error it is, please post an update in this thread, because I'm genuinely curious to know what is causing the problem for you.

**Reply by FHBlah** [reply]
Well, I got Firebase to initialize successfully. As I was retracing my steps, I found that modifications I had made to **Info.plist** were missing. The [TestApp](https://github.com/firebase/quickstart-cpp/tree/main/gma/testapp) **readme.md** file has a link to [instructions](https://developers.google.com/admob/ios/quick-start#manual_download) on adding the `GADApplicationIdentifier` and `SKAdNetworkItems` keys. I had completed the steps but when I double-checked my work, the entries were gone.

In the end, I found that modifying **Info.plist** from within my project doesn't seem to work. It was only once I'd double-clicked the file in Finder, modified it, and saved it that the changes persisted after an Axmol build. I can modify my source files from within my project but not **Info.plist**? Strange. RH, Maybe you have some insights on this?


>So, perhaps you shouldn't use a while loop, and instead use an if condition...

I'm aware that my code renders my app completely unresponsive while it attempts to initialize Firebase. This was just brute force polling to find out if I could get Firebase working. Now that I have, I can write something more UI friendly. Thanks again, RH.

**Reply by rh101** [reply]
> In the end, I found that modifying **Info.plist** from within my project doesn't seem to work. It was only once I'd double-clicked the file in Finder, modified it, and saved it that the changes persisted after an Axmol build. I can modify my source files from within my project but not **Info.plist**? Strange. RH, Maybe you have some insights on this?

At a guess, when generating the xcode project via cmake (`or axmol build` etc.), the `Info.plist` file that is in the `proj.ios_mac\ios\targets\ios` is copied across to the build folder.  So, when the project is opened in Xcode, it's actually using the `Info.plist` that is in the build folder, and not the one that is in the original `proj.ios_mac\ios\targets\ios` folder.  

Once again, this is a complete guess, but you can always check if there is a `Info.plist` file in the build folder too, which would be the temporary copy of the original.

If that isn't the case, and the `Info.plist` that is opened up in Xcode is in fact the original, then at the moment I do not know would be causing it not to save the updates.

Anyhow, to ensure everything works, all edits need to be done to the original `proj.ios_mac\ios\targets\ios\Info.plist`, or via entries in `CMakeLists.txt`.

EDIT: Just verified it, so yes, the `Info.plist` is being copied across to the build folder, and that is the one you see inside Xcode. For example: 

`./build_arm64/bin/myapp/Debug/myapp.app/Contents/Info.plist`

If you edit the `Info.plist` file that you see within Xcode, then you will see the changes reflected in the `Info.plist` within the path listed above.

**Reply by FHBlah** [reply]
Remember when I said that I had gotten Firebase to initialize successfully? I had actually accidentally modified my code to exit the loop even though the initialization status was still `kFutureStatusPending`. Embarrassed, I kept searching for a solution but haven't found one as of yet. As this isn't really an Axmol issue, i've posted a [question](https://stackoverflow.com/questions/78474748/when-i-initialize-my-firebaseapp-why-is-the-resulting-futures-status-forever) on Stack Overflow. I'll let you know if I ever get it figured out. Cheers!

**Reply by FHBlah** [reply]
Update: The folks at Stack Overflow suggested I try using callback functions rather than polling the initialization status. This is a good workaround. I've revised my code to use callbacks and was able to load test ads from Admob. Thanks All.
